### ISSUE TITLE
EBCDIC test failures

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#4130]
An x64 EBCDIC build of the current master fails two tests: the integer stack sub-test and several of the password sub-tests.  A build without the `-DCHARSET_EBCDIC -DCHARSET_EBCDIC_TEST` is fine.

> ./config -Werror -debug --strict-warnings no-shared -DCHARSET_EBCDIC -DCHARSET_EBCDIC_TEST

```
Test Summary Report
-------------------
../test/recipes/02-test_stack.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../test/recipes/20-test_passwd.t (Wstat: 5888 Tests: 25 Failed: 23)
  Failed tests:  2-5, 7-25
  Non-zero exit status: 23
```

The outputs from the two tests are:
```
../test/recipes/02-test_stack.t ... 
1..1
    # Subtest: ./stack_test
    1..4
    # ERROR: (int) 'sk_sint_find_ex(s, &finds[i].value) == finds[i].ex' failed @ test/stack_test.c:144
    # [2] compared to [1]
    # INFO: 
    # int sorted find_ex present 4
    not ok 1 - test_int_stack
    ok 2 - test_uchar_stack
    ok 3 - test_SS_stack
    ok 4 - test_SU_stack
../util/shlib_wrap.sh ./stack_test => 1
not ok 1 - running stack_test

#   Failed test 'running stack_test'
#   at ../test/testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
```

and

```
../test/recipes/20-test_passwd.t .. 
1..25
../util/shlib_wrap.sh ../apps/openssl passwd password => 0
ok 1 - crypt password with random salt
passwd: Option unknown option -1
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -1 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 117.
not ok 2 - BSD style MD5 password with random salt

#   Failed test 'BSD style MD5 password with random salt'
#   at ../test/recipes/20-test_passwd.t line 84.
passwd: Option unknown option -apr1
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -apr1 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 117.
not ok 3 - Apache style MD5 password with random salt

#   Failed test 'Apache style MD5 password with random salt'
#   at ../test/recipes/20-test_passwd.t line 86.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 117.
not ok 4 - SHA256 password with random salt

#   Failed test 'SHA256 password with random salt'
#   at ../test/recipes/20-test_passwd.t line 88.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 117.
not ok 5 - Apache SHA512 password with random salt

#   Failed test 'Apache SHA512 password with random salt'
#   at ../test/recipes/20-test_passwd.t line 90.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xx password => 0
ok 6 - crypt password with salt xx
passwd: Option unknown option -1
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xxxxxxxx -1 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 7 - BSD style MD5 password with salt xxxxxxxx

#   Failed test 'BSD style MD5 password with salt xxxxxxxx'
#   at ../test/recipes/20-test_passwd.t line 95.
passwd: Option unknown option -apr1
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xxxxxxxx -apr1 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 8 - Apache style MD5 password with salt xxxxxxxx

#   Failed test 'Apache style MD5 password with salt xxxxxxxx'
#   at ../test/recipes/20-test_passwd.t line 97.
passwd: Option unknown option -aixmd5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xxxxxxxx -aixmd5 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 9 - AIX style MD5 password with salt xxxxxxxx

#   Failed test 'AIX style MD5 password with salt xxxxxxxx'
#   at ../test/recipes/20-test_passwd.t line 99.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xxxxxxxxxxxxxxxx -5 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 10 - SHA256 password with salt xxxxxxxxxxxxxxxx

#   Failed test 'SHA256 password with salt xxxxxxxxxxxxxxxx'
#   at ../test/recipes/20-test_passwd.t line 101.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -salt xxxxxxxxxxxxxxxx -6 password => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 11 - SHA512 password with salt xxxxxxxxxxxxxxxx

#   Failed test 'SHA512 password with salt xxxxxxxxxxxxxxxx'
#   at ../test/recipes/20-test_passwd.t line 103.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt saltstring 'Hello world!' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 12 - SHA256 password with salt saltstring

#   Failed test 'SHA256 password with salt saltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=10000$saltstringsaltstring' 'Hello world!' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 13 - SHA256 password with salt rounds=10000$saltstringsaltstring

#   Failed test 'SHA256 password with salt rounds=10000$saltstringsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=5000$toolongsaltstring' 'This is just a test' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 14 - SHA256 password with salt rounds=5000$toolongsaltstring

#   Failed test 'SHA256 password with salt rounds=5000$toolongsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=1400$anotherlongsaltstring' 'a very much longer text to encrypt.  This one even stretches over morethan one line.' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 15 - SHA256 password with salt rounds=1400$anotherlongsaltstring

#   Failed test 'SHA256 password with salt rounds=1400$anotherlongsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=77777$short' 'we have a short salt string but not a short password' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 16 - SHA256 password with salt rounds=77777$short

#   Failed test 'SHA256 password with salt rounds=77777$short'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=123456$asaltof16chars..' 'a short string' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 17 - SHA256 password with salt rounds=123456$asaltof16chars..

#   Failed test 'SHA256 password with salt rounds=123456$asaltof16chars..'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -5
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -5 -salt 'rounds=10$roundstoolow' 'the minimum number is still observed' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 18 - SHA256 password with salt rounds=10$roundstoolow

#   Failed test 'SHA256 password with salt rounds=10$roundstoolow'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt saltstring 'Hello world!' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 19 - SHA512 password with salt saltstring

#   Failed test 'SHA512 password with salt saltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=10000$saltstringsaltstring' 'Hello world!' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 20 - SHA512 password with salt rounds=10000$saltstringsaltstring

#   Failed test 'SHA512 password with salt rounds=10000$saltstringsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=5000$toolongsaltstring' 'This is just a test' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 21 - SHA512 password with salt rounds=5000$toolongsaltstring

#   Failed test 'SHA512 password with salt rounds=5000$toolongsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=1400$anotherlongsaltstring' 'a very much longer text to encrypt.  This one even stretches over morethan one line.' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 22 - SHA512 password with salt rounds=1400$anotherlongsaltstring

#   Failed test 'SHA512 password with salt rounds=1400$anotherlongsaltstring'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=77777$short' 'we have a short salt string but not a short password' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 23 - SHA512 password with salt rounds=77777$short

#   Failed test 'SHA512 password with salt rounds=77777$short'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=123456$asaltof16chars..' 'a short string' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 24 - SHA512 password with salt rounds=123456$asaltof16chars..

#   Failed test 'SHA512 password with salt rounds=123456$asaltof16chars..'
#   at ../test/recipes/20-test_passwd.t line 107.
passwd: Option unknown option -6
passwd: Use -help for summary.
../util/shlib_wrap.sh ../apps/openssl passwd -6 -salt 'rounds=10$roundstoolow' 'the minimum number is still observed' => 1
Use of uninitialized value in pattern match (m//) at ../test/recipes/20-test_passwd.t line 124.
not ok 25 - SHA512 password with salt rounds=10$roundstoolow

#   Failed test 'SHA512 password with salt rounds=10$roundstoolow'
#   at ../test/recipes/20-test_passwd.t line 107.
# Looks like you failed 23 tests of 25.
Dubious, test returned 23 (wstat 5888, 0x1700)
Failed 23/25 subtests 
```

**Comment by paulidale** [comment]
A Travis test case should be added once the tests pass.  Include a run with the `no-shared -DCHARSET_EBCDIC -DCHARSET_EBCDIC_SET`.

**Comment by levitte** [comment]
I've a fix for the failing passwd test, #4227
